homeassistant/components/geniushub/binary_sensor.py

```diff
@@ -44,9 +47,13 @@ def device_state_attributes(self) -> Dict[str, Any]:
         attrs = {}
         attrs["assigned_zone"] = self._device.data["assignedZones"][0]["name"]
 
-        # pylint: disable=protected-access
-        last_comms = self._device._raw["childValues"]["lastComms"]["val"]
+        attrs["state"] = dict(self._device.data["state"])
```

Comment:
What data are we storing under state? What are the keys?

Reply:
For `self._device.data["state"]`, we are limited to the `slugs` from this [dict](https://github.com/zxdavb/geniushub-client/blob/c8dc109651a3e831f5d6de9bc012f6f4c6c069ca/geniushubclient/const.py#L355-L398) (from the vendor), namely: 
 - `outputOnOff`,
 - `batteryLevel`,
 - `setTemperature`, 
 - `measuredTemperature`, 
 - `luminance`, and
 - `occupancyTrigger`

For `self._device.data["_state"]`, it is [explicitly](https://github.com/zxdavb/geniushub-client/blob/c8dc109651a3e831f5d6de9bc012f6f4c6c069ca/geniushubclient/__init__.py#L693-L697): 
 - `lastComms`, and
 - `setback`

There is a lot of scope to add to this, as there is a lot of raw data per device, but that will be a new client library!

Reply:
Can we format the keys as snakecase? We can use the stringcase library if needed.

Reply:
Wouldn't many of these values fit better as their own sensors?

Reply:
> Can we format the keys as snakecase? We can use the stringcase library if needed.

I _could_ rename them to snake_case.  The reason why I chose not to is because that's the names the vendor uses (based upon JavaScript conventions, I guess), and I wanted to stay aligned to their schema as much as practical.

I can imagine a scenario where some JSON/dict is exported from HA back to the vendor API (e.g. each zone has two schedules), and it needs converting back to camelCase - I would find this awkward & unsatisfactory.

I would prefer to leave it as is, but will change it if you require.

> Wouldn't many of these values fit better as their own sensors?

I put a lot of thought into this, and feel it is about right, I wouldn't like to change it.

Reply:
@MartinHjelmare I have a question about integrating my client library with HA.

I would like to modify the client, so that it does concurrent I/O.  After each `scan_interval`, it fetches 3-4 URLs, which - thus far- it has been doing sequentially.  

Now I'm planning to the following - is that OK, or is there a HA way of doing it?
```python
async def update(self):
    """Update the Hub with its latest state data."""
    if self.api_version == 1:
        self._zones, self._devices, self._issues, self._version = await asyncio.gather(
            *[
                asyncio.ensure_future(self.request("GET", g))
                for g in ["zones", "devices", "issues", "version"]
            ]
        )

    else:
        zones, data_manager, auth = await asyncio.gather(
            *[
                asyncio.ensure_future(self.request("GET", g))
                for g in ["zones", "data_manager", "auth/release"]
            ]
        )

        self._zones = _zones_via_v3_zones(zones)
        self._devices = _devices_via_v3_data_mgr(data_manager)
        self._issues = _issues_via_v3_zones(zones)
        self._version = _version_via_v3_auth(auth)

    await self._update()  # now convert all the raw JSON

```

Reply:
The whole point of Home Assistant is that we create a unified interface for the user to different integrations. We want all state attribute keys to be lower case snake_case.

Reply:
If `self.request` is a coroutine, use `await asyncio.gather(*coros)` directly. Don't use `asyncio.ensure_future`.

If this code should be in Home Assistant, the library should abstract away `GET` and other protocol specific information. The user of the library shouldn't have to know that type of info to use the library.

Reply:
Yes, it's in the library - made change as you suggested & implemented it now, since it has passed CI, I have bumped the client. 

It was the right thing to do - HA startup decreased noticeably (10s of seconds) in the case of the v1 API call.

`snack_case` is done - I used my own code rather than the library (only 2 lines of code).

Sorry, but I did a bit of a refactor, and other changes (e.g. added a new Climate type).